File: ios/Classes/Methods/MessageMethods.swift
import Flutter
import TwilioConversationsClient

// swiftlint:disable type_body_length
class MessageMethods: NSObject, TWCONMessageApi {
    let TAG = "MessageMethods"

    // swiftlint:disable function_body_length
    public func getMediaContentTemporaryUrlConversationSid(
        _ conversationSid: String?,
        messageIndex: NSNumber?,
        completion: @escaping (String?, FlutterError?) -> Void) {
        debug("getMediaContentTemporaryUrl => conversationSid: \(String(describing: conversationSid)), "
                + "messageIndex: \(String(describing: messageIndex))")
        guard let client = SwiftTwilioConversationsPlugin.instance?.client else {
            return completion(
                nil,
                FlutterError(
                    code: "ClientNotInitializedException",
                    message: "Client has not been initialized.",
                    details: nil))
        }

        guard let conversationSid = conversationSid else {
            return completion(
                nil, FlutterError(
                    code: "MissingParameterException",
                    message: "Missing conversationSid",
                    details: nil))
        }

        guard let messageIndex = messageIndex else {
            return completion(
                nil,
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing messageIndex",
                    details: nil))
        }

        client.conversation(
            withSidOrUniqueName: conversationSid,
            completion: { (result: TCHResult, conversation: TCHConversation?) in
                if result.isSuccessful, let conversation = conversation {
                    conversation.message(
                        withIndex: messageIndex,
                        completion: { (result: TCHResult, message: TCHMessage?) in
                        if result.isSuccessful, let message = message {
                            message.getMediaContentTemporaryUrl(completion: { (result: TCHResult, url: String?) in
                                if result.isSuccessful, let url = url {
                                    self.debug("getMediaContentTemporaryUrl => onSuccess: \(url)")
                                    completion(url, nil)
                                } else {
                                    let errorMessage = String(describing: result.error)
                                    self.debug("getMediaContentTemporaryUrl => onError: \(errorMessage)")
                                    completion(
                                        nil,
                                        FlutterError(
                                            code: "TwilioException",
                                            message: "\(result.error?.code)|Error getting mediaContentTemporaryUrl: "
                                            + "\(errorMessage)",
                                            details: nil))
                                }
                            })
                        } else {
                            self.debug("getMediaContentTemporaryUrl => onError: \(String(describing: result.error))")
                            completion(
                                nil,
                                FlutterError(
                                    code: "NotFoundException",
                                    message: "Error locating message with index \(messageIndex) "
                                        + "in conversation \(conversationSid)",
                                    details: nil))
                        }
                    })
                } else {
                    completion(
                        nil,
                        FlutterError(
                            code: "NotFoundException",
                            message: "Error locating conversation \(conversationSid)",
                            details: nil))
                }
        })
    }

    /// getParticipant
    func getParticipantConversationSid(
        _ conversationSid: String?,
        messageIndex: NSNumber?,
        completion: @escaping (TWCONParticipantData?, FlutterError?) -> Void) {
        debug("getParticipant => conversationSid: \(String(describing: conversationSid)), "
                + "messageIndex: \(String(describing: messageIndex))")
        guard let client = SwiftTwilioConversationsPlugin.instance?.client else {
            return completion(
                nil,
                FlutterError(
                    code: "ClientNotInitializedException",
                    message: "Client has not been initialized.",
                    details: nil))
        }

        guard let conversationSid = conversationSid else {
            return completion(
                nil, FlutterError(
                    code: "MissingParameterException",
                    message: "Missing conversationSid",
                    details: nil))
        }

        guard let messageIndex = messageIndex else {
            return completion(
                nil,
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing messageIndex",
                    details: nil))
        }

        client.conversation(
            withSidOrUniqueName: conversationSid,
            completion: { (result: TCHResult, conversation: TCHConversation?) in
                if result.isSuccessful, let conversation = conversation {
                    conversation.message(
                        withIndex: messageIndex,
                        completion: { (result: TCHResult, message: TCHMessage?) in
                        if result.isSuccessful, let message = message {
                            guard let participant = message.participant else {
                                return completion(
                                    nil,
                                    FlutterError(
                                        code: "NotFoundException",
                                        message: "Participant not found for message: \(messageIndex).",
                                        details: nil))
                            }

                            self.debug("getParticipant => onSuccess")
                            let participantData =
                                Mapper.participantToPigeon(participant, conversationSid: conversationSid)
                            return completion(participantData, nil)
                        } else {
                            self.debug("getParticipant => onError: \(String(describing: result.error))")
                            completion(
                                nil,
                                FlutterError(
                                    code: "NotFoundException",
                                    message: "Error getting message at index \(messageIndex) "
                                        + "in conversation \(conversationSid)",
                                    details: nil))
                        }
                    })
                } else {
                    completion(
                        nil,
                        FlutterError(
                            code: "NotFoundException",
                            message: "Error locating conversation \(conversationSid)",
                            details: nil))
                }
        })
    }

    /// setAttributes
    func setAttributesConversationSid(
        _ conversationSid: String?,
        messageIndex: NSNumber?,
        attributes: TWCONAttributesData?,
        completion: @escaping (FlutterError?) -> Void) {
        debug("setAttributes => conversationSid: \(String(describing: conversationSid)), "
                + "messageIndex: \(String(describing: messageIndex))")
        guard let client = SwiftTwilioConversationsPlugin.instance?.client else {
            return completion(
                FlutterError(
                    code: "ClientNotInitializedException",
                    message: "Client has not been initialized.",
                    details: nil))
        }

        guard let conversationSid = conversationSid else {
            return completion(
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing conversationSid",
                    details: nil))
        }

        guard let messageIndex = messageIndex else {
            return completion(
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing messageIndex",
                    details: nil))
        }

        guard let attributesData = attributes else {
            return completion(
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing 'attributes' parameter",
                    details: nil))
        }

        var messageAttributes: TCHJsonAttributes?
        do {
            messageAttributes = try Mapper.pigeonToAttributes(attributesData)
        } catch LocalizedConversionError.invalidData {
            return completion(
                FlutterError(
                    code: "ConversionException",
                    message: "Could not convert \(attributes?.data) to valid TCHJsonAttributes",
                    details: nil)
            )
        } catch {
            return completion(
                FlutterError(
                    code: "ConversionException",
                    message: "\(attributes?.type) is not a valid type for TCHJsonAttributes.",
                    details: nil)
            )
        }

        client.conversation(
            withSidOrUniqueName: conversationSid,
            completion: { (result: TCHResult, conversation: TCHConversation?) in
            if result.isSuccessful, let conversation = conversation {
                self.debug("setAttributes => onSuccess")
                conversation.message(
                    withIndex: messageIndex,
                    completion: { (result: TCHResult, message: TCHMessage?) in
                    if result.isSuccessful, let message = message {
                        message.setAttributes(messageAttributes) { (result: TCHResult) in
                            if result.isSuccessful {
                                self.debug("setAttributes => onSuccess")
                                return completion(nil)
                            } else {
                                self.debug("setAttributes => onError: \(String(describing: result.error))")
                                return completion(
                                    FlutterError(
                                        code: "TwilioException",
                                        message: "\(result.error?.code)|Error setting attributes for message at "
                                            + "index \(messageIndex) in conversation \(conversationSid)",
                                        details: nil))
                            }
                        }
                    } else {
                        self.debug("setAttributes => onError: \(String(describing: result.error))")
                        completion(
                            FlutterError(
                                code: "NotFoundException",
                                message: "Error getting message at index \(messageIndex) "
                                    + "in conversation \(conversationSid)",
                                details: nil))
                    }
                })
            } else {
                let errorMessage = String(describing: result.error)
                self.debug("setAttributes => onError: \(errorMessage)")
                completion(
                    FlutterError(
                        code: "NotFoundException",
                        message: "Error locating conversation \(conversationSid): \(errorMessage)",
                        details: nil))
            }
        })
    }

    /// updateMessageBody
    func updateMessageBodyConversationSid(
        _ conversationSid: String?,
        messageIndex: NSNumber?,
        messageBody: String?,
        completion: @escaping (FlutterError?) -> Void) {
        debug("updateMessageBody => conversationSid: \(String(describing: conversationSid)), "
                + "messageIndex: \(String(describing: messageIndex))")
        guard let client = SwiftTwilioConversationsPlugin.instance?.client else {
            return completion(
                FlutterError(
                    code: "ClientNotInitializedException",
                    message: "Client has not been initialized.",
                    details: nil))
        }

        guard let conversationSid = conversationSid else {
            return completion(
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing conversationSid",
                    details: nil))
        }

        guard let messageIndex = messageIndex else {
            return completion(
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing messageIndex",
                    details: nil))
        }

        guard let messageBody = messageBody else {
            return completion(
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing messageBody",
                    details: nil))
        }

        client.conversation(
            withSidOrUniqueName: conversationSid,
            completion: { (result: TCHResult, conversation: TCHConversation?) in
                if result.isSuccessful, let conversation = conversation {
                    conversation.message(
                        withIndex: messageIndex,
                        completion: { (result: TCHResult, message: TCHMessage?) in
                        if result.isSuccessful, let message = message {
                            message.updateBody(messageBody) { (result: TCHResult) in
                                if result.isSuccessful {
                                    self.debug("updateMessageBody => onSuccess")
                                    return completion(nil)
                                } else {
                                    self.debug("updateMessageBody => onError: \(String(describing: result.error))")
                                    return completion(
                                        FlutterError(
                                            code: "TwilioException",
                                            message: "\(result.error?.code)|Error updating message at index "
                                                + "\(messageIndex) in conversation \(conversationSid)",
                                            details: nil))
                                }
                            }
                        } else {
                            self.debug("updateMessageBody => onError: \(String(describing: result.error))")
                            completion(
                                FlutterError(
                                    code: "NotFoundException",
                                    message: "Error getting message at index \(messageIndex) "
                                        + "in conversation \(conversationSid)",
                                    details: nil))
                        }
                    })
                } else {
                    completion(
                        FlutterError(
                            code: "NotFoundException",
                            message: "Error locating conversation \(conversationSid)",
                            details: nil))
                }
        })
    }

    private func debug(_ msg: String) {
        SwiftTwilioConversationsPlugin.debug("\(TAG)::\(msg)")
    }
}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/MessageMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.twilio.conversations.*
import com.twilio.conversations.extensions.getDetailedDeliveryReceiptList
import com.twilio.conversations.extensions.getMessageByIndex
import com.twilio.util.ErrorInfo
import com.twilio.conversations.extensions.updateMessageBody
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import twilio.flutter.twilio_conversations.Api
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin
import twilio.flutter.twilio_conversations.exceptions.ClientNotInitializedException
import twilio.flutter.twilio_conversations.exceptions.ConversionException
import twilio.flutter.twilio_conversations.exceptions.NotFoundException
import twilio.flutter.twilio_conversations.exceptions.TwilioException
import twilio.flutter.twilio_conversations.listeners.SafeCallbackListener
import twilio.flutter.twilio_conversations.listeners.SafeStatusListener

class MessageMethods : Api.MessageApi {
    private val TAG = "MessageMethods"

    override fun getMediaContentTemporaryUrl(
        conversationSid: String,
        messageIndex: Long,
        result: Api.Result<String>
    ) {
        debug("getMediaContentTemporaryUrl => conversationSid: $conversationSid")
        val client = TwilioConversationsPlugin.client
            ?: return result.error(ClientNotInitializedException("Client is not initialized"))

        client.getConversation(conversationSid, object : SafeCallbackListener<Conversation> {
            override fun onSafeSuccess(item: Conversation) {
                item.getMessageByIndex(messageIndex, object : SafeCallbackListener<Message> {
                    override fun onSafeSuccess(item: Message) {
                        debug("getMediaContentTemporaryUrl => onSuccess: ${item.attachedMedia}")

                        if (item.attachedMedia.isEmpty()) {
                            return result.error(NotFoundException("No media attached to message"))
                        }

                        // TODO: Add support for multiple media
                        val media = item.attachedMedia.first()

                        media.getTemporaryContentUrl(
                            object : SafeCallbackListener<String> {
                                override fun onSafeSuccess(item: String) {
                                    debug("getMediaContentTemporaryUrl => onSuccess: $item")
                                    result.success(item)
                                }

                                override fun onError(errorInfo: ErrorInfo) {
                                    debug("getMediaContentTemporaryUrl => onError: $errorInfo")
                                    result.error(TwilioException(errorInfo.code, errorInfo.message))
                                }
                            }
                        )
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("getMediaContentTemporaryUrl => onError: $errorInfo")
                        result.error(TwilioException(errorInfo.code, errorInfo.message))
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                debug("getMediaContentTemporaryUrl => onError: $errorInfo")
                result.error(TwilioException(errorInfo.code, errorInfo.message))
            }
        })
    }

    override fun getParticipant(
        conversationSid: String,
        messageIndex: Long,
        result: Api.Result<Api.ParticipantData>
    ) {
        debug("getParticipant => conversationSid: $conversationSid messageIndex: $messageIndex")
        val client = TwilioConversationsPlugin.client
            ?: return result.error(ClientNotInitializedException("Client is not initialized"))

        client.getConversation(conversationSid, object : SafeCallbackListener<Conversation> {
            override fun onSafeSuccess(item: Conversation) {
                item.getMessageByIndex(messageIndex, object : SafeCallbackListener<Message> {
                    override fun onSafeSuccess(item: Message) {
                        val participant = item.participant
                            ?: return result.error(NotFoundException("Participant not found for message: $messageIndex."))
                        debug("getParticipant => onSuccess")
                        return result.success(Mapper.participantToPigeon(participant))
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("getParticipant => onError: $errorInfo")
                        result.error(TwilioException(errorInfo.code, errorInfo.message))
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                debug("getParticipant => onError: $errorInfo")
                result.error(TwilioException(errorInfo.code, errorInfo.message))
            }
        })
    }

    override fun setAttributes(
        conversationSid: String,
        messageIndex: Long,
        attributes: Api.AttributesData,
        result: Api.Result<Void>
    ) {
        debug("setAttributes => conversationSid: $conversationSid messageIndex: $messageIndex")
        val client = TwilioConversationsPlugin.client
            ?: return result.error(ClientNotInitializedException("Client is not initialized"))
        val messageAttributes = Mapper.pigeonToAttributes(attributes)
            ?: return result.error(ConversionException("Could not convert $attributes to valid Attributes"))

        client.getConversation(conversationSid, object : SafeCallbackListener<Conversation> {
            override fun onSafeSuccess(item: Conversation) {
                val conversation = item

                conversation.getMessageByIndex(messageIndex, object :
                    SafeCallbackListener<Message> {
                    override fun onSafeSuccess(item: Message) {
                        item.setAttributes(messageAttributes, object : SafeStatusListener {
                            override fun onSafeSuccess() {
                                debug("setAttributes => onSuccess")
                                result.success(null)
                            }

                            override fun onError(errorInfo: ErrorInfo) {
                                debug("setAttributes => onError: $errorInfo")
                                result.error(TwilioException(errorInfo.code, errorInfo.message))
                            }
                        })
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("setAttributes => onError: $errorInfo")
                        result.error(TwilioException(errorInfo.code, errorInfo.message))
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                debug("setAttributes => onError: $errorInfo")
                result.error(TwilioException(errorInfo.code, errorInfo.message))
            }
        })
    }

    override fun updateMessageBody(
        conversationSid: String,
        messageIndex: Long,
        messageBody: String,
        result: Api.Result<Void>
    ) {
        debug("updateMessageBody => conversationSid: $conversationSid messageIndex: $messageIndex")
        val client = TwilioConversationsPlugin.client
            ?: return result.error(ClientNotInitializedException("Client is not initialized"))

        client.getConversation(conversationSid, object : SafeCallbackListener<Conversation> {
            override fun onSafeSuccess(item: Conversation) {
                item.getMessageByIndex(messageIndex, object : SafeCallbackListener<Message> {
                    override fun onSafeSuccess(item: Message) {
                        // TODO: Handle error case
                        GlobalScope.launch { item.updateMessageBody(messageBody) }
                        result.success(null)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("updateMessageBody => onError: $errorInfo")
                        result.error(TwilioException(errorInfo.code, errorInfo.message))
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                debug("updateMessageBody => onError: $errorInfo")
                result.error(TwilioException(errorInfo.code, errorInfo.message))
            }
        })
    }

    fun debug(message: String) {
        TwilioConversationsPlugin.debug("$TAG::$message")
    }

    override fun getAggregatedDeliveryReceipt(
        conversationSid: String,
        messageIndex: Long,
        result: Api.Result<Api.DeliveryReceiptData>
    ) {
        debug("getAggregatedDeliveryReceipt => conversationSid: $conversationSid messageIndex: $messageIndex")
        val client = TwilioConversationsPlugin.client
            ?: return result.error(ClientNotInitializedException("Client is not initialized"))

        client.getConversation(conversationSid, object : SafeCallbackListener<Conversation> {
            override fun onSafeSuccess(item: Conversation) {
                item.getMessageByIndex(messageIndex, object : SafeCallbackListener<Message> {
                    override fun onSafeSuccess(item: Message) {
                        val aggregatedDeliveryReceipt =
                            item.aggregatedDeliveryReceipt ?: return result.error(
                                NotFoundException("AggregatedDeliveryReceipt not found for message: $messageIndex.")
                            )
                        debug("getAggregatedDeliveryReceipt => onSuccess")
                        return result.success(
                            Mapper.aggregatedDeliveryReceiptToPigeon(
                                aggregatedDeliveryReceipt
                            )
                        )
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("getAggregatedDeliveryReceipt => onError: $errorInfo")
                        result.error(TwilioException(errorInfo.code, errorInfo.message))
                    }
                })
            }

            override fun onError(errorInfo: ErrorInfo) {
                debug("getMediaContentTemporaryUrl => onError: $errorInfo")
                result.error(TwilioException(errorInfo.code, errorInfo.message))
            }
        })
    }

    override fun getDetailedDeliveryReceiptList(
        conversationSid: String,
        messageIndex: Long,
        result: Api.Result<List<Api.DetailedDeliveryReceiptData?>>
    ) {
        debug("getDetailedDeliveryReceiptList => conversationSid: $conversationSid messageIndex: $messageIndex")
        var client = TwilioConversationsPlugin.client ?: return result.error(
            ClientNotInitializedException("Client is not initialized")
        )

        client.getConversation(conversationSid, object : SafeCallbackListener<Conversation> {
            override fun onSafeSuccess(item: Conversation) {
                item.getMessageByIndex(messageIndex, object : SafeCallbackListener<Message> {
                    override fun onSafeSuccess(item: Message) {
                        item.getDetailedDeliveryReceiptList(object :
                            SafeCallbackListener<List<DetailedDeliveryReceipt>> {
                            override fun onSafeSuccess(item: List<DetailedDeliveryReceipt>) {
                                debug("getDetailedDeliveryReceiptList => onSuccess")

                                val elements = item.map {
                                    Mapper.detailedDeliveryReceiptToPigeon(it)
                                }

                                return result.success(elements)
                            }

                            override fun onError(errorInfo: ErrorInfo) {
                                debug("getDetailedDeliveryReceiptList => onError: $errorInfo")
                                result.error(TwilioException(errorInfo.code, errorInfo.message))
                            }
                        })
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("getDetailedDeliveryReceiptList => onError: $errorInfo")
                        result.error(TwilioException(errorInfo.code, errorInfo.message))
                    }
                })
            }
        })
    }
}
